### PR TITLE
feat: leaderboard

### PR DESCRIPTION
**Pull request by yruej301**

### CODE REVIEW

File: packages/indexer-client/src/types/IndexerLeaderboardType.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export type IndexerLeaderboardRankType = 'pnl' | 'roi';
```
Comment (frankfka) [inline packages/indexer-client/src/types/IndexerLeaderboardType.ts:1]: Can we leave a comment explaining the difference? pnl is absolute pnl, roi is %

File: packages/indexer-client/src/types/clientTypes.ts [hunk 2]
```diff
@@ -614,3 +615,60 @@ export interface GetIndexerMakerStatisticsResponse {
...
+export interface GetIndexerLeaderboardParams {
+  contestId: number;
+  rankType: IndexerLeaderboardRankType;
+  // Min rank inclusive
+  startCursor?: string;
+  limit?: number;
+}
+
+export interface IndexerLeaderboardParticipant {
+  subaccount: string;
```
Comment (frankfka) [inline packages/indexer-client/src/types/clientTypes.ts:628]: Shouldn't this be processed into `Subaccount`?

File: packages/indexer-client/src/types/serverTypes.ts [hunk 3]
```diff
@@ -360,6 +382,18 @@ export interface IndexerServerMakerStatisticsResponse {
   makers: IndexerServerMaker[];
 }
 
+export interface IndexerServerLeaderboardResponse {
+  positions: IndexerServerLeaderboardPosition[];
+}
+
+export interface IndexerServerLeaderboardRankResponse {
+  position: IndexerServerLeaderboardPosition;
```
Comment (frankfka) [inline packages/indexer-client/src/types/serverTypes.ts:390]: Doesn't this return `null` when not eligible?

File: packages/indexer-client/src/types/clientTypes.ts (hunk 2 above)
Comment (yruej301) [reply packages/indexer-client/src/types/clientTypes.ts:628]: Yup, will fix

File: packages/indexer-client/src/types/serverTypes.ts (hunk 3 above)
Comment (yruej301) [reply packages/indexer-client/src/types/serverTypes.ts:390]: Yup, will fix

File: packages/indexer-client/src/types/IndexerLeaderboardType.ts (hunk 1 above)
Comment (yruej301) [reply packages/indexer-client/src/types/IndexerLeaderboardType.ts:1]: Yup, will do

Comment (yruej301) [reply packages/indexer-client/src/types/IndexerLeaderboardType.ts:2]: addressed!

File: packages/indexer-client/src/types/clientTypes.ts (hunk 2 above)
Comment (yruej301) [reply packages/indexer-client/src/types/clientTypes.ts]: addressed!

File: packages/indexer-client/src/types/serverTypes.ts (hunk 3 above)
Comment (yruej301) [reply packages/indexer-client/src/types/serverTypes.ts]: addressed!

File: packages/indexer-client/src/types/clientTypes.ts [hunk 4]
```diff
@@ -614,3 +615,60 @@ export interface GetIndexerMakerStatisticsResponse {
...
+  participants: IndexerLeaderboardParticipant[];
+}
+
+export interface GetIndexerLeaderboardParticipantParams {
+  contestId: number;
+  subaccount: Subaccount;
+}
+
+export interface GetIndexerLeaderboardParticipantResponse {
+  participant: IndexerLeaderboardParticipant | null;
```
Comment (frankfka) [inline packages/indexer-client/src/types/clientTypes.ts:648]: nit: can we leave a comment saying that this is null if not eligible?

File: packages/indexer-client/src/types/clientTypes.ts [hunk 5]
```diff
@@ -614,3 +615,60 @@ export interface GetIndexerMakerStatisticsResponse {
...
+  startTime: BigDecimal;
+  // End time in seconds
+  endTime: BigDecimal;
+  // Contest duration in seconds; when set to 0, contest duration is [startTime,endTime];
+  // Otherwise, contest runs indefinitely in the interval [lastUpdated - period, lastUpdated] if active;
+  period: BigDecimal;
+  // Last updated time in Seconds
+  lastUpdated: BigDecimal;
+  totalParticipants: BigDecimal;
+  minRequiredAccountValue: BigDecimal;
```
Comment (frankfka) [inline packages/indexer-client/src/types/clientTypes.ts:668]: We should leave a comment indicating that this does not include any decimals (i.e its a float, 250 = min of 250 USDC)

File: packages/indexer-client/src/types/serverModelTypes.ts [hunk 6]
```diff
@@ -285,3 +285,24 @@ export interface IndexerServerMaker {
   address: string;
   data: IndexerServerMakerData[];
 }
+
+export interface IndexerServerLeaderboardPosition {
```
Comment (frankfka) [inline packages/indexer-client/src/types/serverModelTypes.ts:289]: Not for this PR, but can we add account value to be returned here?

**Review by frankfka — APPROVED** [review]
cc @tmns

File: packages/indexer-client/src/types/clientTypes.ts (hunk 5 above)
Comment (yruej301) [reply packages/indexer-client/src/types/clientTypes.ts:668]: sg

File: packages/indexer-client/src/types/clientTypes.ts (hunk 4 above)
Comment (yruej301) [reply packages/indexer-client/src/types/clientTypes.ts:649]: addressed

File: packages/indexer-client/src/types/clientTypes.ts (hunk 5 above)
Comment (yruej301) [reply packages/indexer-client/src/types/clientTypes.ts:670]: addressed

File: packages/indexer-client/src/types/serverModelTypes.ts (hunk 6 above)
Comment (yruej301) [reply packages/indexer-client/src/types/serverModelTypes.ts:289]: @frankfka is this a must have? we currently don't keep track of account value in this context.

Comment (frankfka) [reply packages/indexer-client/src/types/serverModelTypes.ts:289]: Hmm Conor asked for it - but I thought for ROI calc, you would need account value at last update time?

Comment (yruej301) [reply packages/indexer-client/src/types/serverModelTypes.ts:289]: @frankfka this is addressed

**Review by frankfka — APPROVED** [review]

File: packages/indexer-client/src/types/clientTypes.ts [hunk 7]
```diff
@@ -614,3 +615,64 @@ export interface GetIndexerMakerStatisticsResponse {
...
+  percentRoi: BigDecimal;
+  roiRank: BigDecimal;
+  // Float indicating the ending account value at the time the snapshot was taken i.e: at updateTime
+  accountValue: BigDecimal;
+  // Seconds
+  updateTime: BigDecimal;
+}
+
+export interface GetIndexerLeaderboardResponse {
+  participants: IndexerLeaderboardParticipant[];
```
Comment (tmns) [inline packages/indexer-client/src/types/clientTypes.ts:641]: Hey @yruej301, I'm currently working on the frontend piece of this. 

Should there be some kind of `meta.nextCursor` or `nextPageParam` returned here as well to help with pagination?

Comment (frankfka) [reply packages/indexer-client/src/types/clientTypes.ts:641]: Are we able to derive this with the query params of startCursor & limit?

Comment (frankfka) [reply packages/indexer-client/src/types/clientTypes.ts:641]: Actually, we typically have additional processing to support pagination in `IndexerClient` (i.e. adding `WithPaginationMeta`). @yruej301 are you able to add this for this query as well?

Comment (yruej301) [reply packages/indexer-client/src/types/clientTypes.ts:641]: yup, i will add

**Comment by tmns** [comment]
Hey folks would it be possible to merge this and cut a release? I ran up against issues today linking the SDK locally, started getting the following error response when attempting to fetch (any of the new endpoints):

```
422 (Unprocessable Content), IndexerBaseClient.ts:824

Failed to deserialize the JSON body into the target type: unknown variant `leaderboard_contests`, expected one of `liquidation_feed`, `funding_rate`, `funding_rates`, `candlesticks`, `products`, `product_snapshots`, `events`, `orders`, `summary`, `account_snapshots`, `matches`, `rewards`, `maker_statistics`, `subaccounts`, `price`, `perp_prices`, `oracle_price`, `usdc_price`, `linked_signer_rate_limit`, `linked_signers`, `market_snapshots`, `referral_code`, `interest_and_funding`, `vrtx_merkle_proofs`, `arb_merkle_proofs`, `arb_rewards`, `signatures`, `taker_rewards`, `initial_drop_conditions`, `blitz_points`, `blast_points`, `blitz_points_leaderboard` at line 1 column 23
```
It's the error returned from Axios [here](https://github.com/vertex-protocol/vertex-typescript-sdk/blob/feat/leaderboard/packages/indexer-client/src/IndexerBaseClient.ts#L824), but it wasn't breaking before and I don't think anything's changed 🤔 So I'm guessing something has just gotten really confused on my end. Tried clearing caches and all that, no dice.

**Comment by frankfka** [comment]
@tmns Should be `0.34.0`!

**Pull request merged**